src/ReductionMsg.chpl

```diff
@@ -611,11 +619,39 @@ module ReductionMsg
       return res;
     }
     
-    proc segMean(values:[] ?t, segments:[?D] int): [D] real {
+    proc segMean(values:[] ?t, segments:[?D] int, skipNan=false): [D] real {
       var res: [D] real;
       if (D.size == 0) { return res; }
-      var sums = segSum(values, segments);
-      var counts = segCount(segments, values.size);
+      var sums;
+      var counts;
+      if (isFloatType(t) && skipNan) {
+        // count cumulative nans over all values
+        var cumnans = isnan(values):int;
+        cumnans = + scan cumnans;
+        
+        // find cumulative nans at segment boundaries
+        var segnans: [segments.domain] int;
+        forall si in segments.domain {
+          if si == segments.domain.high {
+              segnans[si] = cumnans[cumnans.domain.high];
+          } else {
+              segnans[si] = cumnans[segments[si+1]-1];
+          }
+        }
+        
+        // take diffs of adjacent segments to find nan count in each segment
+        var nancounts: [segments.domain] int;
+        nancounts[0] = segnans[0];
+        nancounts[1..(segnans.domain.size-1)] = segnans[segnans.domain.interior(segnans.domain.size-1)] - segnans[segnans.domain.interior(-(segnans.domain.size-1))];
```

Comment:
Is there a reason for using the explicit range in the LHS expression? I'm worried that it will trigger some bad access patterns, although @bradcray or @ronawho would know better. I think if we use another interior expression, i.e. `nancounts[nancounts.domain.interior(-(nancounts.domain.size-1))] = ...`, the domains should do the right thing with respect to locality.

Reply:
I'm embarrassed at how quickly my memory has become rusty w.r.t. slicing, but I'm pretty sure that slicing a distributed array with a range will do the correct thing w.r.t. locality (and will continue to in the future).  Recall that other cases where we've tried to move away from slicing with a range were (I believe) ones in which the range described a completely/mostly local set of indices, such that using a default / non-distributed domain rather than a range resulted in performance improvements (?).

A more straightforward way to write this (if I'm reading it correctly) would be to inherit the bounds using:

```chapel
nancounts[1..] = segnans[..segnans.domain.high-1];
```

Seeing Bill's other comment, we may want to make this more index-neutral, though, and use:

```chapel
nancounts[segnans.domain.low+1..] = segnans[..segnans.domain.high-1];
```

(did I translate that correctly?)

That said, I also feel that `segments.domain` and `segnans.domain` seem long-winded.  If I'm reading the code correctly, I think that these domain expressions are the same as `D`?  (which is arguably a bit terse, but...).  If so, and it didn't have any negative performance impacts (I don't think it should), I might use:

```chapel
nancounts[D.low+1..] = segnans[..D.high-1];
```


The following are some other ways I wrote it before arriving at the above:

```chapel
const shiftSize = D.size - 1;
nancounts[D#shiftSize] = nancounts[D#(-shiftSize)];   // copy last n-1 elements into the first n-1
```

Note that `#n` says "give me the first n things" if `n` is positive and "...the final n things" if it's negative.

Thinking about it a moment more, we also ought to be able to write this as:

```chapel
const shiftSize = D.size - 1;
nancounts#shiftSize = nancounts#(-shiftSize);   // copy last n-1 elements into the first n-1
```

though I have a lot less experience applying `#` to arrays (and maybe none to distributed arrays), so am less confident in it (though it _ought_ to do the right thing if implemented correctly).

Reply:
Ah, that makes sense. Based on Brad's comments, I would go with the following: 

```chapel
nancounts[D.low+1..] = segnans[D.low+1..] - segnans[..D.high-1];
```

By the way, that `D#shiftSize` pattern is super cool. I did not know about that.

Reply:
Alright, think I think that I have these changes in now. Thanks for the suggestions guys, what you have suggested obviously looks much cleaner than what I have before.